src/listen/ldap_sync/proto_ldap_sync_ldap.h

```diff
@@ -0,0 +1,125 @@
+} sync_phases_t;
+
+/** State of an individual sync
+ */
+struct sync_state_s {
+	fr_rb_node_t			node;		//!< Entry in the tree of nodes.
+
+	fr_ldap_connection_t 		*conn;		//!< Connection the sync is running on.
+
+	sync_config_t const		*config;	//!< Configuration for this sync
+
+	int				msgid;		//!< The unique identifier for this sync session.
+
+	uint8_t				*cookie;	//!< Opaque cookie, used to resume synchronisation.
+
+	sync_phases_t			phase;		//!< Phase this sync is in.
+
+	fr_dlist_head_t			*filter;	//!< Parsed filter to be applied on the network side
+							//!< before passing packets to the worker.
+							//!< Predominantly to overcome Active Directory's lack
+							//!< of filtering in persistent searches.
+};
+
+typedef struct sync_state_s sync_state_t;
+
+/** Types of LDAP messages relevant to LDAP sync
+ */
+static fr_table_num_sorted_t const sync_ldap_msg_table[] = {
+	{ L("intermediateResponse"),	LDAP_RES_INTERMEDIATE		},
+	{ L("searchRes"),		LDAP_RES_SEARCH_RESULT		},
+	{ L("searchResEntry"),		LDAP_RES_SEARCH_ENTRY		},
+	{ L("searchResReference"),	LDAP_RES_SEARCH_REFERENCE	}
+};
+static size_t const sync_ldap_msg_table_len = NUM_ELEMENTS(sync_ldap_msg_table);
+
+typedef struct {
+	CONF_SECTION			*cs;			//!< our configuration
+
+	proto_ldap_sync_t		*parent;		//!< The module that spawned us.
+
+	fr_ldap_config_t		handle_config;		//!< Connection configuration instance
+
+	sync_config_t			**sync_config;		//!< DNs and filters to monitor.
+
+	char const			*tls_random_file;	//!< Path to the random file if /dev/random
+								//!< and /dev/urandom are unavailable.
+
+	uint32_t			ldap_debug;		//!< Debug flag for the SDK.
+} proto_ldap_sync_ldap_t;
+
+typedef struct {
+	char const			*name;			//!< socket name
+	proto_ldap_sync_ldap_t const	*inst;			//!< instance data
+
+	fr_event_list_t			*el;			//!< Network side event list.
+	fr_network_t			*nr;			//!< Network handler.
+	fr_listen_t			*parent;		//!< master IO handler.
+	fr_listen_t			*li;			//!< Our listener.
+
+	fr_event_timer_t const		*conn_retry_ev;		//!< When to retry re-establishing the conn.
+
+	fr_connection_t			*conn;			//!< Our connection to the LDAP directory.
+} proto_ldap_sync_ldap_thread_t;
+
+/** Tracking structure for connections requiring refresh
+ */
+struct sync_refresh_packet_s {
+	sync_state_t			*sync;			//!< Sync requiring refresh
+
+	uint8_t				*refresh_cookie;	//!< Cookie provided by the server for the refresh.
+};
+
+typedef struct sync_refresh_packet_s sync_refresh_packet_t;
+
+extern fr_table_num_sorted_t const sync_op_table[];
+extern size_t sync_op_table_len;
+
+int8_t sync_state_cmp(void const *one, void const *two);
+
+sync_state_t *sync_state_alloc(TALLOC_CTX *ctx, fr_ldap_connection_t *conn, sync_config_t const *config);
+
+int ldap_sync_cookie_store(sync_state_t *sync, uint8_t const *cookie, bool refresh);
+
+int ldap_sync_entry_send(sync_state_t *sync, uint8_t const uuid[SYNC_UUID_LENGTH], struct berval *orig_dn,
```

Comment:
`uint8_t const uuid[static SYNC_UUID_LENGTH]` to actually attempt to enforce the requirement

Reply:
Only RFC4533 servers provide a UUID, so for the others we have to be able to pass NULL in here.